plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java

```diff
@@ -1,13 +1,223 @@
+          ServerCore.findRuntimeType(AppEngineStandardFacet.DEFAULT_RUNTIME_ID);
+      if (appEngineRuntimeType == null) {
+        throw new NullPointerException("Could not find " + AppEngineStandardFacet.DEFAULT_RUNTIME_NAME + " runtime type");
+      }
+
+      IRuntimeWorkingCopy appEngineRuntimeWorkingCopy
+          = appEngineRuntimeType.createRuntime(null, monitor);
+      CloudSdk cloudSdk = new CloudSdk.Builder().build();
+      if (cloudSdk != null) {
+        java.nio.file.Path sdkLocation = cloudSdk.getJavaAppEngineSdkPath();
+        if (sdkLocation != null) {
+          IPath sdkPath = Path.fromOSString(sdkLocation.toAbsolutePath().toString());
+          appEngineRuntimeWorkingCopy.setLocation(sdkPath);
+        }
+      }
+
+      org.eclipse.wst.server.core.IRuntime appEngineServerRuntime =
+          appEngineRuntimeWorkingCopy.save(true, monitor);
+      IRuntime appEngineFacetRuntime = FacetUtil.getRuntime(appEngineServerRuntime);
+      if (appEngineFacetRuntime == null) {
+        throw new NullPointerException("Could not locate App Engine facet runtime");
+      }
+
+      project.addTargetedRuntime(appEngineFacetRuntime, monitor);
+      project.setPrimaryRuntime(appEngineFacetRuntime, monitor);
+    }
+  }
+
+  /**
+   * Installs Java 1.7 facet if it doesn't already exist in <code>factedProject</code>
+   */
+  private static void installJavaFacet(IFacetedProject facetedProject, IProgressMonitor monitor)
+      throws CoreException {
+    if (facetedProject.hasProjectFacet(JavaFacet.VERSION_1_7)) {
+      return;
+    }
+
+    JavaFacetInstallConfig javaConfig = new JavaFacetInstallConfig();
+    List<IPath> sourcePaths = new ArrayList<>();
+    sourcePaths.add(new Path("src/main/java"));
+    sourcePaths.add(new Path("src/test/java"));
+    javaConfig.setSourceFolders(sourcePaths);
+    facetedProject.installProjectFacet(JavaFacet.VERSION_1_7, javaConfig, monitor);
+  }
+
+  /**
+   * Installs Dynamic Web Module 2.5 facet if it doesn't already exits in <code>factedProject</code>
+   */
+  private static void installWebFacet(IFacetedProject facetedProject, IProgressMonitor monitor)
+      throws CoreException {
+    if (facetedProject.hasProjectFacet(WebFacetUtils.WEB_25)) {
+      return;
+    }
+
+    IDataModel webModel = DataModelFactory.createDataModel(new WebFacetInstallDataModelProvider());
+    webModel.setBooleanProperty(IJ2EEModuleFacetInstallDataModelProperties.ADD_TO_EAR, false);
+    webModel.setBooleanProperty(IJ2EEFacetInstallDataModelProperties.GENERATE_DD, false);
+    webModel.setBooleanProperty(IWebFacetInstallDataModelProperties.INSTALL_WEB_LIBRARY, false);
+    webModel.setStringProperty(IWebFacetInstallDataModelProperties.CONFIG_FOLDER, "src/main/webapp");
+    facetedProject.installProjectFacet(WebFacetUtils.WEB_25, webModel, monitor);
+  }
+
+  // TODO: find a more general form of this method
+  private static org.eclipse.wst.server.core.IRuntime[] getAppEngineRuntime() {
```

Comment:
If it can return more than one runtime, I'd call it `getAppEngineRuntimes` (plural)